src/main/java/com/sarality/sync/APISyncExecutor.java

```diff
@@ -41,56 +48,78 @@ public void execute() {
     execute(null, null);
   }
 
-  public void execute(TaskProgressListener<SyncProgressCount> progressListener,
+  public boolean execute(TaskProgressListener<SyncProgressCount> progressListener,
       TaskCompletionListener<SyncProgressCount> completionListener) {
     this.progressListener = progressListener;
+    hasErrors = false;
+    syncErrorList = new ArrayList<>();
 
     int pageCount = fetcher.init();
 
-
     List<T> sourceDataList = fetcher.fetchNext();
+    // TODO (@Satya) check if there were errors in fetcher and add to error list.
+
     int currentPage = 0;
     int progressCount = 0;
     int itemsOnPage = 0;
 
     while (sourceDataList != null) {
       List<S> collatedList = collator.collate(sourceDataList);
+      // TODO (@Satya) check if there were errors in collator and add to error list.
       itemsOnPage = collatedList.size();
 
       updateProgress(new SyncProgressCount(++currentPage, pageCount, progressCount, itemsOnPage));
 
       for (S data : collatedList) {
-        R request = requestGenerator.generateSyncRequest(data);
+        R request = null;
+        try {
+          request = requestGenerator.generateSyncRequest(data);
+          // TODO (@Satya) check if there were errors and add them to the error list.
+        } catch (Throwable t) {
+          hasErrors = true;
+          addError(new SyncErrorData(APISyncErrorLocation.API_SYNC_EXECUTE,
```

Comment:
When we do the refactoring we may want to use a separate class called SyncErrorCollector which basically has this call of addErrors and hasErrors. The same class can then used by Fetcher, Collator etc either by composition of inheritance, so that errors can be gathered in a the same fashion everywhere

Reply:
got it.